### ISSUE TITLE
Buck daemon event bus encountered an error

### ISSUE THREAD

**Issue by benbrittain** [facebook/buck2#189]
I'm trying to get RBE working, and besides taking longer than I'd expect (but that might be the `buildbarn` setup, unknown yet), it dies before completing a compilation.

```
Jobs completed: 95. Time elapsed: 228.6s. Cache hits: 75%. Commands: 4 (cached: 3, remote: 0, local: 1)
Command failed: Buck daemon event bus encountered an error, the root cause (if available) is displayed above this message.

Caused by:
    0: status: Unknown, message: "error reading a body from connection: stream closed because of a broken pipe", details: [], metadata: MetadataMap { headers: {} }
    1: error reading a body from connection: stream closed because of a broken pipe
    2: stream closed because of a broken pipe
    
 ```
 
I suspect what's happening here is I'm running out of RAM, everything slows to a crawl then I get the error message and `top` indicates the same. The machine I'm on currently only has 16Gb of RAM. 

Is there a way to control the parallelism?

 
Additionally, I tried calling `buck2 rage` since that seemed like an error reporting mechanism 

```
Invocation: (type a number between 0 and 0) 0
Selected invocation at Fri Apr 21 14:48:32 2023 -04:00

2023-04-21T18:52:56.994947Z  WARN buck2_client::commands::rage: Couldn't send rage results to scribe, rage ID `b4325f0f-e40a-467c-bd95-db93ed52823d`
Command failed: Failed to spawn pastry
```
Which is leading me to think it's an internal tool only? Is that right?

**Comment by stepancheg** [comment]
> Is that right?

We should either patch `rage` to generate text data or hide it in open source version.

**Comment by stepancheg** [comment]
CC @iguridi who has implemented the `rage` command.

**Comment by alexlian** [comment]
> Is there a way to control the parallelism?

-j <n>

> although I'm fairly certain linking takes place locally only?

We have links happening remote and locally, largely driven by how much data is being moved IIRC. Guessing @krallin might know more here.

**Comment by krallin** [comment]
Yeah the input size can control where we run, though that’s an option that
I don’t think our PSS examples use (that’s a field on CommandExecutorConfig)

> I suspect what's happening here is I'm running out of RAM, everything
slows to a crawl then I get the error message and top indicates the same.
The machine I'm on currently only has 16Gb of RAM.

And what’s using all this RAM? buck? Something else (linker?)

If too many links are running concurrently then the best approach is
probably to change the link weight, which IIRC is a C++ toolchain field.

**Comment by alexlian** [comment]
We've done some recent work on improving scheduling under constrained resources that probably address the main memory issue here.

@blackm00n pointed me at the following to give useful links and configuration (one needs cgroups v2 on linux for this).

The commit of the improvement is here https://github.com/facebook/buck2/commit/93d31ae98c99437b74f11fb5b70d25294da98e39

Examples are in the e2e tests here
https://github.com/facebook/buck2/commit/979f541840f47f8e7142688ae1df1ce83b5ae417
and
https://github.com/facebook/buck2/commit/b39ce5506e32f44627945c349591e173bdb0d33b

Required configs are:
https://github.com/facebook/buck2/blob/b39ce5506e32f44627945c349591e173bdb0d33b/tests/core/executor/test_hybrid_execution_resource_control_data/.buckconfig#L8-L10

Will close, though please re-open if not a solution for you.
